File: src/main/java/org/lanternpowered/server/config/user/UserStorage.java
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.config.user;

import org.spongepowered.api.profile.GameProfile;

import java.util.Collection;
import java.util.Optional;
import java.util.UUID;

public interface UserStorage<T> {

    /**
     * Gets a collection with all the entries.
     * 
     * @return the entries
     */
    Collection<T> getEntries();

    /**
     * Gets the user entry for the specified unique id.
     * 
     * @param uniqueId the unique id
     * @return the entry if present, otherwise {@link Optional#empty()}
     */
    Optional<T> getEntryByUUID(UUID uniqueId);

    /**
     * Gets the user entry for the specified username.
     * 
     * @param username the username
     * @return the entry if present, otherwise {@link Optional#empty()}
     */
    Optional<T> getEntryByName(String username);

    /**
     * Gets the user entry for the specified game profile.
     * 
     * @param gameProfile the game profile
     * @return the entry if present, otherwise {@link Optional#empty()}
     */
    Optional<T> getEntryByProfile(GameProfile gameProfile);

    /**
     * Adds the op entry and replaces any present ones.
     * 
     * @param entry the entry
     */
    void addEntry(T entry);

    /**
     * Removes the op entry for the specified player unique id.
     * 
     * @param uniqueId the unique id
     * @return whether a entry was removed
     */
    boolean removeEntry(UUID uniqueId);

}

File: src/main/kotlin/org/lanternpowered/api/service/user/UserStorage.kt
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.api.service.user

import org.lanternpowered.api.data.persistence.DataContainer
import org.lanternpowered.api.data.persistence.DataView
import org.lanternpowered.api.util.Identifiable

/**
 * Represents a storage of data for a specific user.
 *
 * The statistic data is present in the path
 * `DataQuery.of("Statistics")`. The data is represented
 * as a map where the key is the id of the statistic and
 * the value is the value of the statistic.
 */
interface UserStorage : Identifiable {

    /**
     * Whether data related to the user exists.
     */
    val exists: Boolean

    /**
     * Loads a [DataContainer] with all user data.
     */
    fun load(): DataContainer

    /**
     * Saves a [DataView] with all user data.
     */
    fun save(data: DataView)

    /**
     * Deletes all data related to the user.
     */
    fun delete(): Boolean
}
